src/config-import.js

```diff
@@ -0,0 +1,271 @@
+import yauzl from 'yauzl-promise'
+import { validate, valueSchemas } from '@mapeo/schema'
+import { json, buffer } from 'node:stream/consumers'
+import path from 'node:path'
+
+// Throw error if a zipfile contains more than 10,000 entries
+const MAX_ENTRIES = 10_000
+
+/**
+ * @typedef {Omit<yauzl.Entry, 'fileName'> & { filename: string }} Entry
+ * `@types/yauzl-promise` has the wrong name for this key. We should
+ * eventually submit this patch upstream.
+ */
+/**
+ * @typedef {{
+ *   presets: { [id: string]: unknown }
+ *   fields:  { [id: string]: unknown }
+ * }} PresetsFile
+ */
+/**
+ * @typedef {Parameters<import('./icon-api.js').IconApi['create']>[0]} IconData
+ */
+
+/**
+ * @param {string} configPath
+ */
+export async function readConfig(configPath) {
+  /** @type {Error[]} */
+  const warnings = []
+
+  const zip = await yauzl.open(configPath)
+  if (zip.entryCount > MAX_ENTRIES) {
+    throw new Error(`Zip file contains too many entries. Max is ${MAX_ENTRIES}`)
+  }
+  /** @type {undefined | Entry} */
+  let presetsEntry
+  for await (const entry of zip) {
+    if (entry.filename === 'presets.json') {
+      presetsEntry = entry
+      break
+    }
+  }
+  if (!presetsEntry) {
+    throw new Error('Zip file does not contain presets.json')
+  }
+  const presetsFile = await json(await presetsEntry.openReadStream())
+  validatePresetsFile(presetsFile)
+
+  return {
+    get errors() {
+      if (warnings.length === 0) return null
+      return warnings
+    },
+
+    async close() {
+      zip.close()
+    },
+
+    /**
+     * @returns {AsyncIterable<IconData>}
+     */
+    async *icons() {
+      /** @type {IconData | undefined} */
+      let icon
+
+      const zip = await yauzl.open(configPath)
+      const entries = await zip.readEntries(MAX_ENTRIES)
+      for (const entry of entries) {
+        if (!entry.filename.match(/^icons\/([^/]+)$/)) continue
+        const buf = await buffer(await entry.openReadStream())
+        const iconFilename = entry.filename.replace(/^icons\//, '')
+        try {
+          const { name, variant } = parseIcon(iconFilename, buf)
+          if (!icon) {
+            icon = {
+              name,
+              variants: [variant],
+            }
+          } else if (icon.name === name) {
+            icon.variants.push(variant)
+          } else {
+            yield icon
+            icon = {
+              name,
+              variants: [variant],
+            }
+          }
+        } catch (err) {
+          warnings.push(err)
+        }
+      }
+      if (icon) {
+        yield icon
+      }
+    },
+
+    /**
+     * @returns {Iterable<{ name: string, value: import('@mapeo/schema').FieldValue }>}
+     */
+    *fields() {
+      const { fields } = presetsFile
+      for (const [name, field] of Object.entries(fields)) {
+        if (!isRecord(field) && !hasOwn(field, 'key')) {
```

Comment:
Should this be an `||`?

```suggestion
        if (!isRecord(field) || !hasOwn(field, 'key')) {
```

Reply:
oops, yeah I missed this and I incidentally this fixed a couple of type errors I had with `field`